nextmv/output.py

```diff
@@ -0,0 +1,392 @@
+                )
+
+            solution = output.solution
+
+        serialized = json.dumps(
+            {
+                "options": options,
+                "solution": solution,
+                "statistics": statistics,
+            },
+            indent=2,
+        )
+
+        if path is None or path == "":
+            print(serialized, file=sys.stdout)
+            return
+
+        with open(path, "w", encoding="utf-8") as file:
+            file.write(serialized + "\n")
+
+    def _write_archive(
+        output: Output,
+        options: Dict[str, Any],
+        statistics: Dict[str, Any],
+        path: Optional[str] = None,
+    ) -> None:
+        dir_path = "output"
+        if path is not None and path != "":
+            if os.path.isfile(path):
+                raise ValueError(f"The path refers to an existing file: {path}")
+
+            dir_path = path
+
+        if not os.path.exists(dir_path):
+            os.makedirs(dir_path)
+
+        serialized = json.dumps(
+            {
+                "options": options,
+                "statistics": statistics,
+            },
+            indent=2,
+        )
+        print(serialized, file=sys.stdout)
+
+        if output.solution is None:
+            return
+
+        for file_name, data in output.solution.items():
+            file_path = os.path.join(dir_path, f"{file_name}.csv")
+            with open(file_path, "w", encoding="utf-8", newline="") as file:
+                writer = csv.DictWriter(file, fieldnames=data[0].keys())
+                writer.writeheader()
+                writer.writerows(data)
+
+    # Callback functions for writing the output data.
+    FILE_WRITERS = {
+        OutputFormat.JSON: _write_json,
+        OutputFormat.CSV_ARCHIVE: _write_archive,
+    }
+
+    def write(self, output: Output, path: Optional[str] = None, *args, **kwargs) -> None:
+        """
+        Write the `output` to the local filesystem. Consider the following for
+        the `path` parameter, depending on the `Output.output_format`:
+
+        - `OutputFormat.JSON`: the `path` is the file where the JSON data will
+            be written. If empty or `None`, the data will be written to stdout.
+        - `OutputFormat.CSV_ARCHIVE`: the `path` is the directory where the CSV
+            files will be written. If empty or `None`, the data will be written
+            to a directory named `output` under the current working directory.
+            The `Output.options` and `Output.statistics` will be written to
+            stdout.
+
+        Parameters
+        ----------
+        output : Output
+            Output data to write.
+        path : str
+            Path to write the output data to.
+        """
+
+        # If the user forgot to reset stdout after redirecting it, we need to
+        # do it here to avoid unexpected behavior.
+        if stdout_redirected:
+            reset_stdout()
```

Comment:
Why do we need to reset stdout redirection here? Aren't there cases where we write to a file instead of stdout? :thinking:

Reply:
This is done so if the user wants to write to stdout, and was using our `redirect_stdout` to redirect the solver chatter to stderr, and forgot to reset stdout to its original state, we do it for them.

Reply:
Otherwise instead of printing the result to stdout they end up printing the result to stderr.